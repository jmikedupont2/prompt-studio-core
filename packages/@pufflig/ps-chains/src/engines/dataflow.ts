import { nodes } from "@pufflig/ps-nodes";
import { NodeType, nodes as nodeConfigMap } from "@pufflig/ps-nodes-config";
import { Param, ParamValue } from "@pufflig/ps-types";
import { Chain, ChainEdge, NodeState } from "../types";

function paramToDefaults(params: Param[]) {
  const defaults: Record<string, ParamValue> = {};
  for (const param of params) {
    defaults[param.id] = param.defaultValue;
  }
  return defaults;
}

function applyDefaultInputs(nodeState: NodeState | undefined, nodeType: NodeType) {
  const nodeConfig = nodeConfigMap[nodeType];
  const inputs = [...nodeConfig.inputs, ...nodeConfig.parameters];
  const defaults = paramToDefaults(inputs);
  return { ...defaults, ...(nodeState?.data || {}) };
}

function getEdgeMap(edge: ChainEdge[]): Record<string, string> {
  const res: Record<string, string> = {};
  for (const e of edge) {
    res[e.source_handle] = e.target_handle;
  }
  return res;
}

function mapOutputToInput(output: Record<string, ParamValue>, map: Record<string, string>) {
  const res: Record<string, ParamValue> = {};
  for (const [key, value] of Object.entries(map)) {
    res[value] = output[key];
  }
  return res;
}

/**
 * This function updates the input state of a node and executes the node on the new input.
 * Read all target nodes and update their input state.
 * Repeat until there are no more target nodes.
 *
 * @param chain current definition and state of the chain
 * @param nodeId id of the node to run
 * @param input input to update the node inputs with
 * @param onNodeStateUpdate callback that triggers when a node input is updated
 */
export async function runFromNode(
  chain: Chain,
  nodeId: string,
  input: Record<string, ParamValue>,
  onNodeStateUpdate: (id: string, state: object) => void
) {
  // avoid hanging on loops by keeping track of the nodes that have been visited
  // current implementation allows a node to be visited only as many times as it has inputs
  const visitedNodes: Record<string, number> = {};
  let chainState: Record<string, NodeState> = chain.state;

  async function _runFromNode(
    chain: Chain,
    nodeId: string,
    input: Record<string, ParamValue>,
    onNodeStateUpdate: (id: string, state: object) => void,
    isChild = false
  ) {
    const nodeDefinition = chain.definition.nodes.find((n) => n.id === nodeId);
    if (!nodeDefinition) {
      throw new Error(`Definition for node ${nodeId} not found`);
    }

    // a node can only be visited as many times as it has connected inputs
    const inputEdges = chain.definition.edges.filter((e) => e.target === nodeId);
    const numInputs = inputEdges.length;

    if (isChild && (visitedNodes[nodeId] || 0) > numInputs) {
      return;
    }

    visitedNodes[nodeId] = visitedNodes[nodeId] ? visitedNodes[nodeId] + 1 : 1;

    const nodeType = nodeDefinition.type;
    const nodeState = chainState[nodeId];
    const newInput = { ...applyDefaultInputs(nodeState, nodeType), ...input };
    const newState = { ...nodeState, data: newInput };

    let newChainState = { ...chainState, [nodeId]: newState };
    chainState = newChainState;

    onNodeStateUpdate(nodeId, newState);

    // do not run nodes in the chain that have autorun=false
    // except for the root node
    if (isChild) {
      if (nodeDefinition.autorun === false) {
        return;
      }
    }

    // get all target nodes
    const targetNodes = chain.definition.edges.filter((e) => e.source === nodeId);
    if (targetNodes.length === 0) {
      return;
    }

    // run the node we updated the input of
    const res = await nodes[nodeType].execute(newInput);
    const editedKeys = Object.keys(res);

    // ignore target nodes that are not connected to the values that have been edited
    const targetNodeIds = targetNodes.filter((e) => editedKeys.includes(e.source_handle)).map((e) => e.target);
    if (targetNodeIds.length === 0) {
      return;
    }

    // update the state of all target nodes
    for (const targetNodeId of targetNodeIds) {
      // edges between this node and the target node
      const edges = chain.definition.edges.filter((e) => e.source === nodeId && e.target === targetNodeId);
      const edgeMap = getEdgeMap(edges);
      const mappedInput = mapOutputToInput(res, edgeMap);
      await _runFromNode(chain, targetNodeId, mappedInput, onNodeStateUpdate, true);
    }
  }

  await _runFromNode(chain, nodeId, input, onNodeStateUpdate);

  return chainState;
}
