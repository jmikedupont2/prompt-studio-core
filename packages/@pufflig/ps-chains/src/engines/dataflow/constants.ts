import { NextNode, Node, ParamValueMap } from "@pufflig/ps-types";

export const delimiterStart = "${{ps:ref:" as const;
export const delimiterEnd = "}}" as const;
export const executionPrefix = "exec:";

export const identity = (i: ParamValueMap, _: Partial<ParamValueMap>) => i;
export const getDefaultTargets =
  (node: Node) => (_input: ParamValueMap, _prev: ParamValueMap, results: ParamValueMap) => {
    if (!node.execution?.outputs?.[0]?.id) return [];
    return [
      {
        execSource: node.execution.outputs[0].id,
        inputs: results,
      } as NextNode,
    ];
  };
