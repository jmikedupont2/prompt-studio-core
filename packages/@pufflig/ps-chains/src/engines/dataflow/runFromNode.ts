import { ParamValue } from "@pufflig/ps-types";
import { Chain, NodeState, RunOptions } from "../../types";
import { resolveVariables } from "./utils/resolveVariables";
import { applyDefaultInputs, getEdgeMap, mapOutputToInput } from "./utils/utils";

/**
 * This function updates the input state of a node and executes the node on the new input.
 * Read all target nodes and update their input state.
 * Repeat until there are no more target nodes.
 *
 * @param nodeId id of the node to run
 * @param input input to update the node inputs with, empty object if existing inputs should be used
 * @param chain current definition and state of the chain
 * @param runOptions functions to call when a node state is updated or an error occurs
 */
export async function runFromNode(
  nodeId: string,
  input: Record<string, ParamValue>,
  chain: Chain,
  runOptions?: RunOptions
) {
  // avoid hanging on loops by keeping track of the nodes that have been visited
  // current implementation allows a node to be visited only as many times as it has inputs
  const visitedNodes: Record<string, number> = {};
  let chainState: Record<string, NodeState> = chain.state;

  async function _runFromNode(
    nodeId: string,
    input: Record<string, ParamValue>,
    chain: Chain,
    runOptions?: RunOptions,
    isChild = false
  ) {
    const nodeConfig = chain.definition.nodes[nodeId];
    const nodeDefinition = chain.nodeTypes[nodeConfig?.type];
    if (!nodeConfig || !nodeDefinition) {
      throw new Error(`Definition for node ${nodeId} not found`);
    }

    // parse the input
    const nodeState = chainState[nodeId];
    const prevInput = nodeState?.input;
    const parsedInput = await nodeDefinition.parseInput(input, prevInput);

    // a node can only be visited as many times as it has connected inputs
    const inputEdges = Object.values(chain.definition.edges).filter((e) => e.target === nodeId);
    const numInputs = inputEdges.length;

    if (isChild && (visitedNodes[nodeId] || 0) > numInputs) {
      return;
    }

    visitedNodes[nodeId] = visitedNodes[nodeId] ? visitedNodes[nodeId] + 1 : 1;

    const newInput = { ...applyDefaultInputs(prevInput, nodeDefinition), ...parsedInput };
    const newState: NodeState = { ...nodeState, input: newInput };

    let newChainState = { ...chainState, [nodeId]: newState };
    chainState = newChainState;

    runOptions?.onNodeInputUpdate?.(nodeId, newState);

    // do not run nodes in the chain that have autorun=false
    // except for the root node
    if (isChild) {
      if (nodeConfig.autorun === false) {
        return;
      }
    }

    // get all target nodes of the node we updated the input of
    // avoid running the node if it has no target nodes
    const targetNodes = Object.values(chain.definition.edges).filter((e) => e.source === nodeId);
    if (targetNodes.length === 0) {
      return;
    }

    let resolvedInput = newInput;
    try {
      resolvedInput = await resolveVariables(newInput, runOptions?.resolveReferences || (async (i) => i));
    } catch (error) {
      runOptions?.onNodeRunError?.(nodeId, error as Error);
      return;
    }

    // run the node we updated the input of
    let res;
    try {
      // resolve references in the input
      res = await nodeDefinition.execute(resolvedInput, prevInput);
      runOptions?.onNodeRunComplete?.(nodeId, res);
      // break the chain if the node returns null
      if (res === null) return;
    } catch (error) {
      // if running the node fails, stop the chain
      runOptions?.onNodeRunError?.(nodeId, error as Error);
      return;
    }

    const editedKeys = Object.keys(res);

    // ignore target nodes that are not connected to the values that have been edited
    const targetNodeIds = targetNodes.filter((e) => editedKeys.includes(e.sourceHandle)).map((e) => e.target);

    // update the state of all target nodes
    for (const targetNodeId of targetNodeIds) {
      // edges between this node and the target node
      const edges = Object.values(chain.definition.edges).filter(
        (e) => e.source === nodeId && e.target === targetNodeId
      );
      const edgeMap = getEdgeMap(edges);
      const mappedInput = mapOutputToInput(res, edgeMap);
      await _runFromNode(targetNodeId, mappedInput, chain, runOptions, true);
    }
  }

  await _runFromNode(nodeId, input, chain, runOptions);

  return chainState;
}
