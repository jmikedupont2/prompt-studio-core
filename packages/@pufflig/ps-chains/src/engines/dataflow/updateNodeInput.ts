import { ParamValue } from "@pufflig/ps-types";
import { Chain, NodeState, RunOptions } from "../../types";
import { applyDefaultInputs } from "./utils/utils";
import _ from "lodash";

/**
 * This function updates the input state of a node without executing the chain.
 *
 * @param nodeId id of the node to run
 * @param input input to update the node inputs with, empty object if existing inputs should be used
 * @param chain current definition and state of the chain
 * @param runOptions
 */
export async function updateNodeInput(
  nodeId: string,
  input: Record<string, ParamValue>,
  chain: Chain,
  runOptions?: RunOptions
) {
  const chainState: Record<string, NodeState> = { ...chain.state };
  const nodeConfig = chain.definition.nodes[nodeId];
  const nodeDefinition = chain.nodeTypes[nodeConfig?.type];

  if (!nodeConfig || !nodeDefinition) {
    throw new Error(`Definition for node ${nodeId} not found`);
  }

  // parse the input
  const nodeState = chainState[nodeId];
  const prevInput = nodeState?.input;
  const parsedInput = await nodeDefinition.parseInput(input, prevInput);

  // compile the new state of the node
  const newInput = { ...applyDefaultInputs(prevInput, nodeDefinition), ...parsedInput };
  const newState: NodeState = { ...nodeState, input: newInput };

  runOptions?.onNodeInputUpdate?.(nodeId, newState);

  const newChainState = _.merge(chainState, { [nodeId]: newState });
  return newChainState;
}
