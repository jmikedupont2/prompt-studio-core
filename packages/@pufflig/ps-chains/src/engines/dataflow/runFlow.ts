import { ParamValue } from "@pufflig/ps-types";
import pino from "pino";
import { Flow, RunOptions } from "../../types";
import { executionPrefix, getDefaultTargets, identity } from "./constants";
import { updateNodeInput } from "./updateNodeInput";
import { getReachableNodes } from "./utils/getReachableNodes";
import { resolveVariables } from "./utils/resolveVariables";
import { getEdgeMap, mapOutputToInput } from "./utils/utils";

const logger = pino();

/**
 * This function updates the input state of a node and executes the node with the new input.
 * Read all target nodes and update their input state. Repeat until there are no more data target nodes.
 * Nodes that are run on execution are run separately.
 *
 * @param flow
 * @param nodeId
 * @param input
 * @param runOptions
 * @returns
 */
export async function runFlow(flow: Flow, nodeId: string, input: Record<string, ParamValue>, runOptions?: RunOptions) {
  logger.level = runOptions?.logLevel || "error";

  const newState = { ...flow.state };

  // track nodes that are reachable from the current node
  const reachableNodes = getReachableNodes(nodeId, flow.definition);

  // avoid hanging on loops by keeping track of the edges that have been visited
  const visitedEdges: string[] = [];
  // track the number of times a node has been run
  const runs: Record<string, number> = {};

  const run_flow_recursive = async (nodeId: string, input: Record<string, ParamValue>) => {
    const previousState = { ...newState[nodeId] };
    const nodeConfig = flow.definition.nodes[nodeId];
    const node = flow.nodeTypes[nodeConfig?.type];

    // update node state with the input values
    const updatedFlow = { ...flow, state: newState };
    const newNodeState = await updateNodeInput(updatedFlow, nodeId, input, runOptions);
    newState[nodeId] = { ...newNodeState };
    logger.debug({ nodeId, input, newState: newState[nodeId] }, "Updated node input");

    const targets = Object.values(flow.definition.edges).filter((edge) => edge.source === nodeId);

    // guard against infinite loops
    const isAlreadyRun = (runs[nodeId] || 0) > 0;

    if (isAlreadyRun) {
      logger.debug({ nodeId }, "Node already run");
      return;
    }

    // break execution if there are no targets
    if (targets.length === 0) {
      logger.debug("No targets found");
      return;
    }

    // if the node is executable, only run it if the parent is complete
    const parentNodeId = Object.values(flow.definition.edges).find(
      (e) => e.target === nodeId && e.targetHandle.startsWith(executionPrefix)
    )?.source;
    const isParentComplete = parentNodeId ? runs[parentNodeId] > 0 : true;
    const isExecutableNode = !!node.execution?.inputs.length;

    if (!isParentComplete && isExecutableNode) {
      logger.debug("Parent node is not complete, skip execution");
      return;
    }

    // resolve variables to use during the run
    const newInput = newState[nodeId]?.input;
    let resolvedInput: Record<string, ParamValue> = {};
    try {
      logger.debug("Resolving node variables");
      resolvedInput = await resolveVariables(newInput, runOptions?.resolveReferences || (async (i) => i));
    } catch (error) {
      logger.error({ error, nodeId }, "Error resolving variables");
      runOptions?.onNodeRunError?.(nodeId, error as Error);
      return;
    }

    const execute = node.execute || identity;
    const getTargets = node.getTargets || getDefaultTargets(node);

    // run the node
    let result: Record<string, ParamValue> | null = null;
    try {
      // resolve references in the input
      const nodeOptions = { prevInput: previousState.input, globals: runOptions?.globals || {} };
      result = await execute(resolvedInput, nodeOptions);
      runOptions?.onNodeRunComplete?.(nodeId, result);
      logger.debug({ nodeId, result }, "Node run complete");
      // track the run
      runs[nodeId] = runs[nodeId] ? runs[nodeId] + 1 : 1;
      // break the chain if the node returns null
      if (result === null) return;
    } catch (error) {
      logger.error({ error, nodeId }, "Error running node");
      // if running the node fails, stop the chain
      runOptions?.onNodeRunError?.(nodeId, error as Error);
      return;
    }

    // ignore target nodes that are not connected to the values that have been edited
    const editedKeys = Object.keys(result);
    const dataTargetIds = targets.filter((e) => editedKeys.includes(e.sourceHandle)).map((e) => e.target);
    logger.debug({ nodeId, dataTargetIds }, "Data targets");

    // update the input of all target nodes
    for (const targetNodeId of dataTargetIds) {
      // data edges between this node and the target node
      const dataEdges = Object.values(flow.definition.edges).filter(
        (e) => e.source === nodeId && e.target === targetNodeId && !e.sourceHandle.startsWith(executionPrefix)
      );
      const edgeMap = getEdgeMap(dataEdges);
      const mappedInput = mapOutputToInput(result, edgeMap);

      // append to visited edges
      visitedEdges.push(...dataEdges.map((e) => e.id));

      // only run the node if all incoming edges have been resolved and ignore parents that are not reachable
      const incomingEdges = Object.values(flow.definition.edges).filter(
        (e) => e.target === targetNodeId && reachableNodes.has(e.source)
      );

      const areIncomingEdgesVisited = incomingEdges.every((e) => visitedEdges.includes(e.id));
      const targetConfig = flow.definition.nodes[targetNodeId];
      const targetDefinition = flow.nodeTypes[targetConfig?.type];
      const targetIsExecutable = !!targetDefinition?.execution?.inputs?.length;
      const shouldRunTarget = areIncomingEdgesVisited && !targetIsExecutable;

      if (shouldRunTarget) {
        logger.debug({ nodeId, targetNodeId, mappedInput }, "Running target node");
        await run_flow_recursive(targetNodeId, mappedInput);
      } else {
        logger.debug({ nodeId, targetNodeId, mappedInput }, "Updating target node inputs");
        const newNodeState = await updateNodeInput(updatedFlow, targetNodeId, mappedInput, runOptions);
        newState[targetNodeId] = { ...newNodeState };
      }
    }

    // run the execution target
    if (runOptions?.mode === "dataflow") return;

    // define the execution order
    const executionOrder = await getTargets(resolvedInput, result, { prevInput: previousState.input });
    const executionTargets = targets.filter(
      (edge) => edge.sourceHandle.startsWith(executionPrefix) && edge.source === nodeId
    );

    logger.debug({ executions: executionOrder }, "Execution Order");

    for (const execution of executionOrder) {
      visitedEdges.push(...executionTargets.map((e) => e.id));
      const targetId = executionTargets.find((e) => e.sourceHandle === execution.execSource)?.target;

      if (targetId) {
        // override the infinite loop guard
        runs[targetId] = runs[targetId] ? runs[targetId] - 1 : 0;

        // data edges between this node and the target node
        const dataEdges = Object.values(flow.definition.edges).filter(
          (e) => e.source === nodeId && e.target === targetId && !e.sourceHandle.startsWith(executionPrefix)
        );

        const edgeMap = getEdgeMap(dataEdges);
        const mappedInput = edgeMap ? mapOutputToInput(execution.inputs, edgeMap) : {};

        logger.debug({ nodeId, targetId, mappedInput }, "Running executable node");
        await run_flow_recursive(targetId, mappedInput);
      } else {
        logger.debug({ executionTargets, executionSource: execution.execSource }, "No target found for source");
      }
    }
  };

  await run_flow_recursive(nodeId, input);
  return newState;
}
