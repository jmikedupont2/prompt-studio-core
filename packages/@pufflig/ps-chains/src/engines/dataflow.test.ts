import { runFromNode } from "./dataflow";
import {
  autorunExample,
  autorunRootExample,
  missingStates,
  multiInput,
  simpleChain,
  simpleLoop,
  singleNode,
} from "./mocks";

test("set input for single node in the editor", async () => {
  const onNodeStateUpdate = jest.fn();
  const res = await runFromNode(singleNode, "1", { template: "Hello World" }, onNodeStateUpdate);
  expect(onNodeStateUpdate).toHaveBeenCalledTimes(1);
  expect(res).toMatchSnapshot();
});

test("set input for several nodes in the editor", async () => {
  const onNodeStateUpdate = jest.fn();
  const res = await runFromNode(simpleChain, "1", { template: "Hello {{World}}" }, onNodeStateUpdate);
  expect(onNodeStateUpdate).toHaveBeenCalledTimes(3);
  expect(res).toMatchSnapshot();
});

test("ignore autorun=false flag on root node", async () => {
  const onNodeStateUpdate = jest.fn();
  const res = await runFromNode(autorunRootExample, "1", { template: "Hello {{World}}" }, onNodeStateUpdate);
  expect(onNodeStateUpdate).toHaveBeenCalledTimes(3);
  expect(res).toMatchSnapshot();
});

test("nodes with autorun=false down the chain do not update the inputs of their targets", async () => {
  const onNodeStateUpdate = jest.fn();
  const res = await runFromNode(autorunExample, "1", { template: "Hello {{World}}" }, onNodeStateUpdate);
  expect(onNodeStateUpdate).toHaveBeenCalledTimes(2);
  expect(res).toMatchSnapshot();
});

test("avoid hanging on loops", async () => {
  const onNodeStateUpdate = jest.fn();
  const res = await runFromNode(simpleLoop, "1", { template: "Hello {{World}}" }, onNodeStateUpdate);
  expect(onNodeStateUpdate).toHaveBeenCalledTimes(2);
  expect(res).toMatchSnapshot();
});

test("can handle missing states", async () => {
  const onNodeStateUpdate = jest.fn();
  const res = await runFromNode(
    missingStates,
    "197375bb-c777-4be5-a423-6d5618e2200f",
    { template: "Hello {{World}}" },
    onNodeStateUpdate
  );
  expect(onNodeStateUpdate).toHaveBeenCalledTimes(4);
  expect(res).toMatchSnapshot();
});

test("can handle input from multiple origins", async () => {
  const onNodeStateUpdate = jest.fn();
  const res = await runFromNode(multiInput, "1", { template: "Hello {{World}}" }, onNodeStateUpdate);
  expect(onNodeStateUpdate).toHaveBeenCalledTimes(4);
  expect(res).toMatchSnapshot();
});

test("update array variables", async () => {
  const onNodeStateUpdate = jest.fn();
  const res = await runFromNode(
    simpleChain,
    "2",
    {
      variables: [
        { id: "world", name: "world", type: "text", description: "", defaultValue: "test1" },
        { id: "hello", name: "hello", type: "text", description: "", defaultValue: "test2" },
      ],
    },
    onNodeStateUpdate
  );
  expect(onNodeStateUpdate).toHaveBeenCalledTimes(2);
  expect(res).toMatchSnapshot();
});
