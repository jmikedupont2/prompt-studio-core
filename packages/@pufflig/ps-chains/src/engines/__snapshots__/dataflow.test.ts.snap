// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`avoid hanging on loops 1`] = `
{
  "1": {
    "data": {
      "template": "Hello {{World}}",
    },
    "status": "idle",
  },
}
`;

exports[`can handle input from multiple origins 1`] = `
{
  "1": {
    "data": {
      "template": "Hello {{World}}",
    },
    "status": "idle",
  },
  "2": {
    "data": {
      "template": "Hello {{World}}",
      "variables": [],
    },
  },
  "3": {
    "data": {
      "completion": "Hello ",
      "prompt": "Hello {{World}}",
    },
  },
}
`;

exports[`can handle missing states 1`] = `
{
  "197375bb-c777-4be5-a423-6d5618e2200f": {
    "data": {
      "template": "Hello {{World}}",
    },
    "status": "idle",
  },
  "2708636f-633c-4c02-abec-093fdc79f4d5": {
    "data": {
      "completion": "",
      "prompt": "Hello ",
    },
  },
  "85c8bc66-cc5a-4400-b4d3-bc119ed1b2f7": {
    "data": {
      "template": "Hello {{World}}",
      "variables": [],
    },
  },
  "c96eeaef-a4bb-4fa6-a32f-b9ad176456ea": {
    "data": {
      "api_key": "\${{SECRET:openai/api_key}}",
      "model": {
        "modelId": "text-davinci-003",
        "parameters": {},
      },
      "prompt": "Hello ",
    },
  },
}
`;

exports[`ignore autorun=false flag on root node 1`] = `
{
  "1": {
    "data": {
      "template": "Hello {{World}}",
    },
    "status": "idle",
  },
  "2": {
    "data": {
      "template": "Hello {{World}}",
      "variables": [
        {
          "defaultValue": "mars",
          "description": "",
          "id": "World",
          "name": "World",
          "type": "text",
        },
      ],
    },
    "status": "idle",
  },
  "3": {
    "data": {
      "completion": "",
      "prompt": "Hello mars",
    },
    "status": "idle",
  },
}
`;

exports[`nodes with autorun=false down the chain do not update the inputs of their targets 1`] = `
{
  "1": {
    "data": {
      "template": "Hello {{World}}",
    },
    "status": "idle",
  },
  "2": {
    "data": {
      "template": "Hello {{World}}",
      "variables": [
        {
          "defaultValue": "mars",
          "description": "",
          "id": "World",
          "name": "World",
          "type": "text",
        },
      ],
    },
    "status": "idle",
  },
  "3": {
    "data": {},
    "status": "idle",
  },
}
`;

exports[`set input for several nodes in the editor 1`] = `
{
  "1": {
    "data": {
      "template": "Hello {{World}}",
    },
    "status": "idle",
  },
  "2": {
    "data": {
      "template": "Hello {{World}}",
      "variables": [
        {
          "defaultValue": "mars",
          "description": "",
          "id": "World",
          "name": "World",
          "type": "text",
        },
      ],
    },
    "status": "idle",
  },
  "3": {
    "data": {
      "completion": "",
      "prompt": "Hello mars",
    },
    "status": "idle",
  },
}
`;

exports[`set input for single node in the editor 1`] = `
{
  "1": {
    "data": {
      "template": "Hello World",
    },
    "status": "idle",
  },
}
`;
